core/services/blockhashstore/feeder.go

```diff
@@ -114,3 +132,114 @@ func (f *Feeder) Run(ctx context.Context) error {
 	f.lastRunBlock = latestBlock
 	return errs
 }
+
+func (f *Feeder) runTrusted(
+	ctx context.Context,
+	latestBlock uint64,
+	fromBlock uint64,
+	blockToRequests map[uint64]map[string]struct{},
+) error {
+	var errs error
+
+	// Iterate through each request block via waitGroup.
+	// For blocks with pending requests, add them to the batch to be stored.
+	// Note: Golang maps sort items in a range randomly, so although the batch size is used
+	// to limit blocks-per-batch, every block has an equal chance of getting picked up
+	// on each run.
+	var batch = make(map[uint64]struct{})
+	for blockKey, unfulfilledReqs := range blockToRequests {
+		f.wgStored.Add(1)
+		var unfulfilled = unfulfilledReqs
+		var block = blockKey
+		go func() {
+			defer f.wgStored.Done()
+			if len(unfulfilled) == 0 {
+				return
+			}
+			if _, ok := f.stored[block]; ok {
+				// Already stored
+				return
+			}
+
+			// Do not store a block if it has been marked as stored; otherwise, store it even
+			// if the RPC call errors, as to be conservative.
+			stored, err := f.bhs.IsStored(ctx, block)
+			if err != nil {
+				f.lggr.Errorw("Failed to check if block is already stored, attempting to store anyway",
+					"err", err,
+					"block", block)
+				errs = multierr.Append(errs, errors.Wrap(err, "checking if stored"))
+			} else if stored {
+				f.lggr.Infow("Blockhash already stored",
+					"block", block, "latestBlock", latestBlock,
+					"unfulfilledReqIDs", LimitReqIDs(unfulfilled, 50))
+				f.stored[block] = struct{}{}
+				return
+			}
+
+			// If there's room, store the block in the batch. Threadsafe.
+			f.batchLock.Lock()
+			if len(batch) < int(f.trustedBHSBatchSize) {
+				batch[block] = struct{}{}
+			}
+			f.batchLock.Unlock()
+		}()
+	}
+
+	// Ensure all blocks are checked before storing the batch.
+	f.wgStored.Wait()
```

Comment:
Kind of wary of this because `ctx` is passed in and it's not clear what the timeout is on it. Should we have aggressive timeouts on the `IsStored` methods (or the network calls in general) so that we can make sure this doesn't hang for a long time?

Reply:
Added a timeout context for the IsStored (1 second based on a local const).